gbpservice/nfp/service_plugins/firewall/nfp_fwaas_plugin.py

```diff
@@ -33,28 +39,53 @@ def _get_routers_for_create_firewall(self, tenant_id, context, firewall):
-            ctx = neutron_context.get_admin_context()
-            routers = l3_plugin.get_routers(ctx)
-            router_ids = [
-                router['id']
-                for router in routers
-                if router['tenant_id'] == tenant_id]
-            # validation can still fail this if there is another fw
-            # which is associated with one of these routers.
-            # Dont validate this condition
-            # self.validate_firewall_routers_not_in_use(context, router_ids)
-            return router_ids
-        else:
-            if not router_ids:
-                # This indicates that user specifies no routers.
-                return []
-            else:
-                # some router(s) provided.
-                # Dont validate this condition
-                # self.validate_firewall_routers_not_in_use(context,
-                # router_ids)
-                return router_ids
+            return tenant_id
+
+    def nfp_create_firewall(self, context, firewall, status=None):
+        fw = firewall['firewall']
+        tenant_id = fw['tenant_id']
+        # distributed routers may required a more complex state machine;
+        # the introduction of a new 'CREATED' state allows this, whilst
+        # keeping a backward compatible behavior of the logical resource.
+        if not status:
+            status = n_const.PENDING_CREATE
+        with context.session.begin(subtransactions=True):
+            self._validate_fw_parameters(context, fw, tenant_id)
+            firewall_db = n_firewall.Firewall(
+                    id=uuidutils.generate_uuid(),
+                    tenant_id=tenant_id,
+                    name=fw['name'],
+                    description=fw['description'],
+                    firewall_policy_id=fw['firewall_policy_id'],
+                    admin_state_up=fw['admin_state_up'],
+                    status=status)
+            context.session.add(firewall_db)
+        return self._make_firewall_dict(firewall_db)
+
+    n_firewall.Firewall_db_mixin.create_firewall = nfp_create_firewall
+
+    def set_routers_for_firewall(self, context, fw):
+        """Sets the routers associated with the fw."""
+        pass
+
+    def get_firewall_routers(self, context, fwid):
+        """Gets all routers associated with a firewall."""
+        fw_rtrs = ['1234567890']
+        return fw_rtrs
+
+    def validate_firewall_routers_not_in_use(
+            self, context, router_ids, fwid=None):
+        """Validate if router-ids not associated with any firewall.
+
+        If any of the router-ids in the list is already associated with
+        a firewall, raise an exception else just return.
+        """
+        pass
+
+    def update_firewall_routers(self, context, fw):
+        """Update the firewall with new routers.
+
+        This involves removing existing router associations and replacing
+        it with the new router associations provided in the update method.
+        """
+        return fw
```

Comment:
Now N-S L3 patch can be added here as monkey patch. Looks like code is different between liberty and mitaka.

For mitaka
from neutron.db import l3_db
def nfp_get_router_floatingip()
"""existing code"""
"""lines 135-141 from N-S extraroute patch"
"""existing Notfound exception"
}
""" Lines 1-127 from extraroute patch"

l3_db.L3_NAT_dbonly_mixin.get_router_for_floatingip = nfp_get_router_for_floatingip